src/layer/wmts.utils.ts

```diff
@@ -0,0 +1,124 @@
+import { RequestConfiguration } from '..';
+import { BBox } from '../bbox';
+import { CRS_EPSG3857, CRS_EPSG4326 } from '../crs';
+import { OgcServiceTypes } from './const';
+import { fetchGetCapabilitiesXml, GetCapabilitiesXmlLayer } from './utils';
+
+const DEGREE_TO_RADIAN = Math.PI / 180;
+const RADIAN_TO_DEGREE = 180 / Math.PI;
+const EQUATOR_RADIUS = 6378137.0;
+const EQUATOR_LEN = 40075016.685578488;
+const MAXEXTENT = 20037508.342789244;
+
+export type GetCapabilitiesWmtsXml = {
+  Capabilities: {
+    'ows:ServiceIdentification': any[][];
+    'ows:ServiceProvider': any[][];
+    'ows:OperationsMetadata': any[][];
+    Contents: [{ Layer: GetCapabilitiesXmlWmtsLayer[] }];
+    ServiceMetadataURL: any[][];
+  };
+};
+
+export type GetCapabilitiesXmlWmtsLayer = {
+  'ows:Title': string[];
+  'ows:Abstract': string[];
+  'ows:WGS84BoundingBox': {}[][];
+  'ows:Identifier': string[];
+  Style: any[];
+  Format: string[];
+  TileMatrixSetLink: any[];
+  ResourceURL: any[];
+};
+
+function toWgs84(xy: number[]): number[] {
+  return [
+    (xy[0] * RADIAN_TO_DEGREE) / EQUATOR_RADIUS,
+    (Math.PI * 0.5 - 2.0 * Math.atan(Math.exp(-xy[1] / EQUATOR_RADIUS))) * RADIAN_TO_DEGREE,
+  ];
+}
+
+function toMercator(xy: number[]): number[] {
+  var xy = [
+    EQUATOR_RADIUS * xy[0] * DEGREE_TO_RADIAN,
+    EQUATOR_RADIUS * Math.log(Math.tan(Math.PI * 0.25 + 0.5 * xy[1] * DEGREE_TO_RADIAN)),
+  ];
+  // if xy value is beyond maxextent (e.g. poles), return maxextent.
+  xy[0] > MAXEXTENT && (xy[0] = MAXEXTENT);
+  xy[0] < -MAXEXTENT && (xy[0] = -MAXEXTENT);
+  xy[1] > MAXEXTENT && (xy[1] = MAXEXTENT);
+  xy[1] < -MAXEXTENT && (xy[1] = -MAXEXTENT);
+  return xy;
+}
+
+function getZoomFromBbox(bbox: BBox): number {
+  if (bbox.crs === CRS_EPSG4326) {
+    const topLeft = toMercator([bbox.minX, bbox.minY]);
+    const bottomRight = toMercator([bbox.maxX, bbox.maxY]);
+    bbox = new BBox(CRS_EPSG3857, topLeft[0], topLeft[1], bottomRight[0], bottomRight[1]);
+  }
+  return Math.max(
+    0,
+    Math.min(19, 1 + Math.floor(Math.log(EQUATOR_LEN / ((bbox.maxX - bbox.minX) * 1.001)) / Math.log(2))),
```

Comment:
I don't really understand this function fully. Does it assume the bbox exactly matches a tile?

Reply:
Currently we can pass a the bbox of a tile, and get the correct X,Y,Z of that tile from a bbox, this will later be extended to support arbitrary bboxes